### ISSUE TITLE
deepseek_train+TPyosoku_seikouritu.py

### ISSUE THREAD

**Issue by akutsuwakutsu** [akutsuwakutsu/.github.io#18]
import pandas as pd
import numpy as np
import lightgbm as lgb
import os

# ========================================
# 設定
# ========================================

NUM_LEAVES = 60
LEARNING_RATE = 0.05
N_ESTIMATORS = 700

TP_LIST = [0.20, 0.30, 0.40, 0.50]   # 20,30,40,50 pips
SL = 0.20                             # 20 pips

TEST_YEARS = list(range(2017, 2026))
TRAIN_YEARS_BACK = 2

# 予測確率のビン（0.02刻み、最後は0.70以上）
BIN_EDGES = [round(0.50 + 0.02*i, 2) for i in range(11)] + [np.inf]
BIN_LABELS = [f"{BIN_EDGES[i]:.2f}-{BIN_EDGES[i+1]:.2f}" for i in range(10)] + ["0.70+"]

# ========================================
# データ読み込み・前処理（前回と同一）
# ========================================

def load_1m_csv(filename):
    print(f"読み込み: {filename}")
    df = pd.read_csv(
        filename,
        header=None,
        names=["Date", "Time", "Open", "High", "Low", "Close", "Volume"]
    )
    df["Datetime"] = pd.to_datetime(df["Date"] + " " + df["Time"])
    df = df.set_index("Datetime")
    df = df[~df.index.duplicated(keep='first')]
    df = df[["Open", "High", "Low", "Close", "Volume"]]
    return df

def make_5m(df_1m):
    df_5m = df_1m.resample("5min").agg({
        "Open": "first",
        "High": "max",
        "Low": "min",
        "Close": "last",
        "Volume": "sum"
    })
    return df_5m.dropna()

def make_features(df_5m):
    x = pd.DataFrame(index=df_5m.index)

    close = df_5m["Close"]
    high = df_5m["High"]
    low = df_5m["Low"]
    open_price = df_5m["Open"]

    x["return_1"] = close.pct_change(1)
    x["return_3"] = close.pct_change(3)
    x["return_6"] = close.pct_change(6)
    x["return_12"] = close.pct_change(12)

    price_range = high - low

    x["range_1"] = price_range.rolling(1).mean()
    x["range_3"] = price_range.rolling(3).mean()
    x["range_6"] = price_range.rolling(6).mean()
    x["range_12"] = price_range.rolling(12).mean()

    ma5 = close.rolling(5).mean()
    ma10 = close.rolling(10).mean()
    ma20 = close.rolling(20).mean()
    x["ma_distance_5"] = close - ma5
    x["ma_distance_10"] = close - ma10
    x["ma_distance_20"] = close - ma20

    hour = df_5m.index.hour + df_5m.index.minute / 60
    x["hour_sin"] = np.sin(2 * np.pi * hour / 24)
    x["hour_cos"] = np.cos(2 * np.pi * hour / 24)

    x["volume_change"] = df_5m["Volume"].pct_change()

    previous_close = close.shift(1)
    tr1 = high - low
    tr2 = (high - previous_close).abs()
    tr3 = (low - previous_close).abs()
    true_range = pd.concat([tr1, tr2, tr3], axis=1).max(axis=1)

    x["atr_6"] = true_range.rolling(6).mean()
    x["atr_12"] = true_range.rolling(12).mean()
    x["atr_24"] = true_range.rolling(24).mean()
    x["range_atr_ratio"] = price_range / x["atr_12"]

    body = (close - open_price).abs()
    x["body_size"] = body
    x["body_range_ratio"] = body / price_range.replace(0, np.nan)

    high_12 = high.rolling(12).max()
    low_12 = low.rolling(12).min()
    high_24 = high.rolling(24).max()
    low_24 = low.rolling(24).min()

    x["distance_from_high_12"] = high_12 - close
    x["distance_from_low_12"] = close - low_12
    x["distance_from_high_24"] = high_24 - close
    x["distance_from_low_24"] = close - low_24

    x["slope_6"] = close - close.shift(6)
    x["slope_12"] = close - close.shift(12)
    x["slope_24"] = close - close.shift(24)

    direction_12 = (close - close.shift(12)).abs()
    volatility_12 = close.diff().abs().rolling(12).sum()
    x["efficiency_ratio_12"] = direction_12 / volatility_12.replace(0, np.nan)

    x = x.replace([np.inf, -np.inf], np.nan)
    return x

def compute_dxy_1m(pair_data):
    common_index = None
    for df in pair_data.values():
        if common_index is None:
            common_index = df.index
        else:
            common_index = common_index.intersection(df.index)

    aligned = {}
    for name, df in pair_data.items():
        aligned[name] = df.reindex(common_index)

    eurusd = aligned["EURUSD"]["Close"]
    usdjpy = aligned["USDJPY"]["Close"]
    gbpusd = aligned["GBPUSD"]["Close"]
    usdcad = aligned["USDCAD"]["Close"]
    usdsek = aligned["USDSEK"]["Close"]
    usdchf = aligned["USDCHF"]["Close"]

    dxy = (
        50.14348112
        * (eurusd ** -0.576)
        * (usdjpy ** 0.136)
        * (gbpusd ** -0.119)
        * (usdcad ** 0.091)
        * (usdsek ** 0.042)
        * (usdchf ** 0.036)
    )
    return pd.DataFrame({"DXY": dxy}, index=common_index)

def make_dxy_features(dxy_1m):
    dxy_5m = pd.DataFrame(index=dxy_1m.resample("5min").last().index)
    dxy_5m["Open"]  = dxy_1m["DXY"].resample("5min").first()
    dxy_5m["High"]  = dxy_1m["DXY"].resample("5min").max()
    dxy_5m["Low"]   = dxy_1m["DXY"].resample("5min").min()
    dxy_5m["Close"] = dxy_1m["DXY"].resample("5min").last()
    dxy_5m = dxy_5m.dropna()

    x = pd.DataFrame(index=dxy_5m.index)
    close = dxy_5m["Close"]
    high = dxy_5m["High"]
    low = dxy_5m["Low"]

    x["dxy_return_1"] = close.pct_change(1)
    x["dxy_return_3"] = close.pct_change(3)
    x["dxy_return_6"] = close.pct_change(6)
    x["dxy_return_12"] = close.pct_change(12)

    previous_close = close.shift(1)
    tr1 = high - low
    tr2 = (high - previous_close).abs()
    tr3 = (low - previous_close).abs()
    true_range = pd.concat([tr1, tr2, tr3], axis=1).max(axis=1)

    x["dxy_atr_6"] = true_range.rolling(6).mean()
    x["dxy_atr_12"] = true_range.rolling(12).mean()
    x["dxy_atr_24"] = true_range.rolling(24).mean()

    x = x.replace([np.inf, -np.inf], np.nan)
    return x

# ========================================
# ラベル作成（TP・SLを引数で指定）
# ========================================

def make_labels(df_5m, df_1m, tp, sl):
    print(f"ラベル作成中... TP={tp}, SL={sl}")
    labels = []
    counts = {"win": 0, "loss": 0, "simultaneous": 0, "no_result": 0, "skip": 0}

    one_min_times = df_1m.index.values
    one_min_high = df_1m["High"].values
    one_min_low = df_1m["Low"].values

    five_min_times = df_5m.index.values
    five_min_open = df_5m["Open"].values

    n = len(df_5m)

    for i in range(n):
        if i == n - 1:
            labels.append(np.nan)
            counts["skip"] += 1
            continue

        entry_time = five_min_times[i + 1]
        entry_price = five_min_open[i + 1]

        entry_idx = np.searchsorted(one_min_times, entry_time)
        if entry_idx >= len(one_min_times):
            labels.append(np.nan)
            counts["skip"] += 1
            continue

        tp_price = entry_price + tp
        sl_price = entry_price - sl

        outcome = None
        for j in range(entry_idx, len(one_min_times)):
            h = one_min_high[j]
            l = one_min_low[j]

            hit_tp = h >= tp_price
            hit_sl = l <= sl_price

            if hit_tp and hit_sl:
                outcome = "simultaneous"
                break
            elif hit_tp:
                outcome = "win"
                break
            elif hit_sl:
                outcome = "loss"
                break

        if outcome == "win":
            labels.append(1)
            counts["win"] += 1
        elif outcome == "loss":
            labels.append(0)
            counts["loss"] += 1
        elif outcome == "simultaneous":
            labels.append(np.nan)
            counts["simultaneous"] += 1
        else:
            labels.append(np.nan)
            counts["no_result"] += 1

    print(f"  内訳: 勝ち={counts['win']}, 負け={counts['loss']}, "
          f"同時={counts['simultaneous']}, No-Result={counts['no_result']}, "
          f"スキップ={counts['skip']}")
    return pd.Series(labels, index=df_5m.index), counts

# ========================================
# 年次データ処理（キャッシュ付き）
# ========================================

cache = {}
label_cache = {}

def load_year_data(year):
    if year in cache:
        return cache[year]

    print(f"\n===== {year}年 データ読み込み・前処理 =====")
    pair_files = {
        "USDJPY": f"USDJPY_{year}_all.csv",
        "EURUSD": f"EURUSD_{year}_all.csv",
        "GBPUSD": f"GBPUSD_{year}_all.csv",
        "USDCAD": f"USDCAD_{year}_all.csv",
        "USDSEK": f"USDSEK_{year}_all.csv",
        "USDCHF": f"USDCHF_{year}_all.csv",
    }

    pair_data = {}
    for name, fname in pair_files.items():
        if not os.path.exists(fname):
            print(f"ファイルが見つかりません: {fname}")
            cache[year] = None
            return None
        pair_data[name] = load_1m_csv(fname)

    dxy_1m = compute_dxy_1m(pair_data)

    df_usdjpy_1m = pair_data["USDJPY"]
    df_usdjpy_5m = make_5m(df_usdjpy_1m)
    features = make_features(df_usdjpy_5m)

    dxy_features = make_dxy_features(dxy_1m)
    features = features.join(dxy_features, how="left")

    data = {
        "df_1m": df_usdjpy_1m,
        "df_5m": df_usdjpy_5m,
        "features": features
    }
    cache[year] = data
    return data

def get_labels(year, tp, sl):
    key = (year, tp)
    if key in label_cache:
        return label_cache[key]

    data = load_year_data(year)
    if data is None:
        label_cache[key] = (None, None)
        return None, None

    labels, counts = make_labels(data["df_5m"], data["df_1m"], tp, sl)
    label_cache[key] = (labels, counts)
    return labels, counts

# ========================================
# メイン処理（確率校正分析）
# ========================================

print("========================================")
print("予測確率と実績成功率の対応関係分析")
print(f"固定パラメータ: num_leaves={NUM_LEAVES}, learning_rate={LEARNING_RATE}, n_estimators={N_ESTIMATORS}")
print(f"TP候補: {TP_LIST} (pips: {[int(tp*100) for tp in TP_LIST]})")
print(f"SL: {SL} ({SL*100:.0f} pips)")
print("========================================")

# 各TPについて、テスト年全体の予測と実績ラベルを集めたDataFrameを保存
# all_pred_actual[tp] = pd.DataFrame({'probability': ..., 'label': ...})
all_pred_actual = {tp: [] for tp in TP_LIST}

for tp in TP_LIST:
    print(f"\n########## TP = {tp} ({int(tp*100)} pips) ##########")

    for test_year in TEST_YEARS:
        train_years = list(range(test_year - TRAIN_YEARS_BACK, test_year))
        print(f"  {train_years}年学習 → {test_year}年テスト")

        # 学習データ準備
        train_features_list = []
        train_labels_list = []
        for year in train_years:
            data = load_year_data(year)
            if data is None:
                continue
            labels, _ = get_labels(year, tp, SL)
            if labels is None:
                continue
            train_features_list.append(data["features"])
            train_labels_list.append(labels)

        if len(train_features_list) == 0:
            continue

        train_features = pd.concat(train_features_list)
        train_labels = pd.concat(train_labels_list)
        train = train_features.copy()
        train["target"] = train_labels
        train = train.dropna()

        FEATURES = list(train_features.columns)
        X_train = train[FEATURES]
        y_train = train["target"].astype(int)

        # テストデータ
        test_data = load_year_data(test_year)
        if test_data is None:
            continue
        X_test = test_data["features"].dropna()
        test_index = X_test.index

        # テスト年のラベル取得
        test_labels, _ = get_labels(test_year, tp, SL)
        if test_labels is None:
            continue

        # モデル学習
        model = lgb.LGBMClassifier(
            objective="binary",
            n_estimators=N_ESTIMATORS,
            learning_rate=LEARNING_RATE,
            num_leaves=NUM_LEAVES,
            random_state=42,
            verbosity=-1
        )
        model.fit(X_train, y_train)

        # 予測
        probabilities = model.predict_proba(X_test)[:, 1]
        predictions = pd.Series(probabilities, index=test_index)

        # 実績ラベルと予測を結合
        # ラベルはすべての5mバーに対して存在するが、X_testのインデックスと揃える
        result_df = pd.DataFrame({
            "probability": predictions,
            "label": test_labels.loc[predictions.index]
        }).dropna(subset=["label"])  # 同時到達・No-Result・スキップは除外

        all_pred_actual[tp].append(result_df)

    # 全体を結合
    if all_pred_actual[tp]:
        all_pred_actual[tp] = pd.concat(all_pred_actual[tp], ignore_index=False)
    else:
        all_pred_actual[tp] = pd.DataFrame(columns=["probability", "label"])

# ========================================
# 予測確率帯別の集計
# ========================================

# 各TPごとの確率帯別集計を表示
for tp in TP_LIST:
    print(f"\n\n===== TP = {tp} ({int(tp*100)} pips) 確率帯別実績 =====")

    df = all_pred_actual[tp]
    if df.empty:
        print("データなし")
        continue

    df["bin"] = pd.cut(df["probability"], bins=BIN_EDGES, labels=BIN_LABELS, right=False)

    summary = df.groupby("bin", observed=False).agg(
        count=("label", "size"),
        wins=("label", "sum"),
        losses=("label", lambda x: (x == 0).sum())
    )

    summary["win_rate"] = summary["wins"] / (summary["wins"] + summary["losses"]) * 100
    summary["profit_pips"] = (summary["wins"] * tp - summary["losses"] * SL) * 100

    # 表示
    print(f"{'確率帯':<12} {'件数':>6} {'勝ち':>6} {'負け':>6} {'実績勝率':>10} {'総損益(pips)':>14}")
    print("-" * 60)
    for idx, row in summary.iterrows():
        print(f"{idx:<12} {int(row['count']):>6} {int(row['wins']):>6} {int(row['losses']):>6} "
              f"{row['win_rate']:>9.2f}% {row['profit_pips']:>+13.1f}")

# ========================================
# 比較表の作成
# ========================================

# 予測確率帯別 実績成功率と件数
print("\n\n========================================")
print("【予測確率帯別 実績成功率】")
print("========================================")

# 各TPの確率帯別集計を取得して結合
summary_dict = {}
for tp in TP_LIST:
    df = all_pred_actual[tp]
    if df.empty:
        continue
    df["bin"] = pd.cut(df["probability"], bins=BIN_EDGES, labels=BIN_LABELS, right=False)
    summary = df.groupby("bin", observed=False).agg(
        count=("label", "size"),
        wins=("label", "sum"),
        losses=("label", lambda x: (x == 0).sum())
    )
    summary["win_rate"] = summary["wins"] / (summary["wins"] + summary["losses"]) * 100
    summary_dict[tp] = summary

# 行を確率帯、列をTPごとの勝率と件数で整理
all_bins = BIN_LABELS
df_winrate = pd.DataFrame(index=all_bins)
df_count = pd.DataFrame(index=all_bins)

for tp in TP_LIST:
    if tp in summary_dict:
        df_winrate[f"TP{int(tp*100)}実績勝率"] = summary_dict[tp]["win_rate"]
        df_count[f"TP{int(tp*100)}件数"] = summary_dict[tp]["count"]
    else:
        df_winrate[f"TP{int(tp*100)}実績勝率"] = np.nan
        df_count[f"TP{int(tp*100)}件数"] = np.nan

# 結合して表示
print("実績勝率（%）:")
print(df_winrate.to_string(float_format=lambda x: f"{x:.2f}"))
print("\n件数:")
print(df_count.to_string(float_format=lambda x: f"{int(x)}"))

# 予測確率帯別 総損益
print("\n\n========================================")
print("【予測確率帯別 総損益】")
print("========================================")

df_profit = pd.DataFrame(index=all_bins)
for tp in TP_LIST:
    if tp in summary_dict:
        profit_pips = (summary_dict[tp]["wins"] * tp - summary_dict[tp]["losses"] * SL) * 100
        df_profit[f"TP{int(tp*100)}損益(pips)"] = profit_pips
    else:
        df_profit[f"TP{int(tp*100)}損益(pips)"] = np.nan

print(df_profit.to_string(float_format=lambda x: f"{x:+.1f}"))

print("\n処理完了")
